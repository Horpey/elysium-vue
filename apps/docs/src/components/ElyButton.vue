<script setup lang="ts">
import type { ButtonHTMLAttributes, FunctionalComponent } from 'vue'
import { RouterLink } from 'vue-router'
import type { ElyButtonSizeKey, ElyButtonThemeKey } from '~/types/ElyButton'

interface Props extends /* @vue-ignore */ ButtonHTMLAttributes {
  type?: 'button' | 'submit' | 'reset'
  theme?: ElyButtonThemeKey
  disabled?: boolean
  leftIcon?: FunctionalComponent
  rightIcon?: FunctionalComponent
  loading?: boolean
  size?: ElyButtonSizeKey
  to?: string
  href?: string
  target?: string
  onClick?: () => void
}

const props = withDefaults(defineProps<Props>(), {
  to: undefined,
  href: undefined,
  target: undefined,
  rel: undefined,
  type: 'button',
  disabled: false,
  theme: 'PRIMARY',
  size: 'MEDIUM',
  onClick: () => {},
})

const isDisabled = computed(() => props.disabled || props.loading)

const buttonClass = computed(() => {
  const baseClass = 'border flex items-center space-x-2 px-5 py-2 text-xs font-light uppercase tracking-wider transition-all duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-opacity-50 shadow-btn'
  let themeClass = ''
  const sizeClass = ''

  switch (props.theme) {
    case 'PRIMARY':
      themeClass = 'border-transparent bg-koromiko-700 text-white focus:ring-koromiko-600'
      break
    case 'SECONDARY':
      themeClass = 'border-transparent bg-aquamarine-700 text-white focus:ring-aquamarine-600'
      break
    case 'TERTIARY':
      themeClass = 'border-transparent bg-codgray-300 text-white focus:ring-codgray-200'
      break
    case 'DESTRUCTIVE':
      themeClass = 'border-transparent bg-rose-700 text-white focus:ring-rose-600'
      break
  }

  const classes = [baseClass, themeClass, sizeClass]

  if (isDisabled.value)
    classes.push('cursor-not-allowed opacity-50')

  return classes.join(' ')
})
</script>

<template>
  <RouterLink
    v-if="to"
    :class="buttonClass"
    :disabled="disabled"
    :to="to"
  >
    <span><slot /></span>
  </RouterLink>

  <a
    v-else-if="href"
    :class="buttonClass"
    :disabled="disabled"
    :href="href"
    :target="target"
  >
    <span><slot /></span>
  </a>

  <button
    v-else
    :class="buttonClass"
    :disabled="disabled"
    @click="onClick"
  >
    <ElyLoader v-if="loading" />
    <span v-if="leftIcon">
      <component
        :is="leftIcon"
        :class="{
          'h-3 w-3': size === 'SMALL',
          'h-4 w-4': size === 'MEDIUM',
          'h-5 w-5': size === 'LARGE',
        }"
      />
    </span>
    <span><slot /></span>
    <span v-if="rightIcon">
      <component
        :is="rightIcon"
        :class="{
          'h-3 w-3': size === 'SMALL',
          'h-4 w-4': size === 'MEDIUM',
          'h-5 w-5': size === 'LARGE',
        }"
      />
    </span>
  </button>
</template>
