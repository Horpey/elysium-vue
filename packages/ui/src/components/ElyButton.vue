<script setup lang="ts">
import type { ButtonHTMLAttributes, FunctionalComponent } from 'vue'
import { RouterLink } from 'vue-router'
import ElyLoader from './ElyLoader.vue'
import type { ElyButtonSizeKey, ElyButtonThemeKey } from '~/types/ElyButton'

interface Props extends /* @vue-ignore */ ButtonHTMLAttributes {
  type?: 'button' | 'submit' | 'reset'
  theme?: ElyButtonThemeKey
  disabled?: boolean
  leftIcon?: FunctionalComponent
  rightIcon?: FunctionalComponent
  loading?: boolean
  size?: ElyButtonSizeKey
  to?: string
  href?: string
  target?: string
  onClick?: () => void
}

const props = withDefaults(defineProps<Props>(), {
  to: undefined,
  href: undefined,
  target: undefined,
  rel: undefined,
  type: 'button',
  disabled: false,
  theme: 'PRIMARY',
  size: 'MEDIUM',
  onClick: () => { },
})

const isDisabled = computed(() => props.disabled || props.loading)

const buttonClass = computed(() => {
  const baseClass = 'border font-light uppercase tracking-wider transition-all duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-opacity-50'

  let themeClass = ''
  let sizeClass = ''

  switch (props.theme) {
    case 'PRIMARY':
      themeClass = 'border-transparent bg-koromiko-700 text-white focus:ring-koromiko-600 hover:bg-koromiko-600 hover:text-white'
      break
    case 'SECONDARY':
      themeClass = 'border-transparent bg-aquamarine-700 text-white focus:ring-aquamarine-600 hover:bg-aquamarine-600 hover:text-white'
      break
    case 'TERTIARY':
      themeClass = 'border-transparent bg-codgray-300 text-white focus:ring-codgray-200 hover:bg-codgray-50 hover:text-white'
      break
    case 'DESTRUCTIVE':
      themeClass = 'border-transparent bg-rose-700 text-white focus:ring-rose-600 hover:bg-rose-600 hover:text-white'
      break
  }

  switch (props.size) {
    case 'SMALL':
      sizeClass = 'text-xs px-3 py-1 h-8'
      break
    case 'MEDIUM':
      sizeClass = 'text-sm px-4 py-2'
      break
    case 'LARGE':
      sizeClass = 'text-base px-6 py-3'
      break
  }

  const classes = [baseClass, themeClass, sizeClass]

  if (isDisabled.value)
    classes.push('cursor-not-allowed opacity-50')

  return classes.join(' ')
})
</script>

<template>
  <RouterLink v-if="to" class="inline-block" :class="buttonClass" :disabled="disabled" :to="to">
    <span
      :class="{
        'flex items-center': loading || leftIcon || rightIcon,
        'space-x-1': size === 'SMALL' && (loading || leftIcon || rightIcon),
        'space-x-2': size === 'MEDIUM' || size === 'LARGE' && (loading || leftIcon || rightIcon),
      }"
    >
      <ElyLoader v-if="loading" :size="size" theme="LIGHT" />
      <span v-if="leftIcon">
        <component
          :is="leftIcon" :class="{
            'h-3 w-3': size === 'SMALL',
            'h-4 w-4': size === 'MEDIUM',
            'h-5 w-5': size === 'LARGE',
          }"
        />
      </span>
      <span>
        <slot />
      </span>
      <span v-if="rightIcon">
        <component
          :is="rightIcon" :class="{
            'h-3 w-3': size === 'SMALL',
            'h-4 w-4': size === 'MEDIUM',
            'h-5 w-5': size === 'LARGE',
          }"
        />
      </span>
    </span>
  </RouterLink>

  <a v-else-if="href" class="inline-block" :class="buttonClass" :disabled="disabled" :href="href" :target="target">
    <span
      :class="{
        'flex items-center': loading || leftIcon || rightIcon,
        'space-x-1': size === 'SMALL' && (loading || leftIcon || rightIcon),
        'space-x-2': size === 'MEDIUM' || size === 'LARGE' && (loading || leftIcon || rightIcon),
      }"
    >
      <ElyLoader v-if="loading" :size="size" theme="LIGHT" />
      <span v-if="leftIcon">
        <component
          :is="leftIcon" :class="{
            'h-3 w-3': size === 'SMALL',
            'h-4 w-4': size === 'MEDIUM',
            'h-5 w-5': size === 'LARGE',
          }"
        />
      </span>
      <span>
        <slot />
      </span>
      <span v-if="rightIcon">
        <component
          :is="rightIcon" :class="{
            'h-3 w-3': size === 'SMALL',
            'h-4 w-4': size === 'MEDIUM',
            'h-5 w-5': size === 'LARGE',
          }"
        />
      </span>
    </span>
  </a>

  <button v-else :class="buttonClass" :disabled="disabled" v-bind="$attrs" :type="type" @click="onClick">
    <span
      :class="{
        'flex items-center': loading || leftIcon || rightIcon,
        'space-x-1': size === 'SMALL' && (loading || leftIcon || rightIcon),
        'space-x-2': size === 'MEDIUM' || size === 'LARGE' && (loading || leftIcon || rightIcon),
      }"
    >
      <ElyLoader v-if="loading" :size="size" theme="LIGHT" />
      <span v-if="leftIcon">
        <component
          :is="leftIcon" :class="{
            'h-3 w-3': size === 'SMALL',
            'h-4 w-4': size === 'MEDIUM',
            'h-5 w-5': size === 'LARGE',
          }"
        />
      </span>
      <span>
        <slot />
      </span>
      <span v-if="rightIcon">
        <component
          :is="rightIcon" :class="{
            'h-3 w-3': size === 'SMALL',
            'h-4 w-4': size === 'MEDIUM',
            'h-5 w-5': size === 'LARGE',
          }"
        />
      </span>
    </span>
  </button>
</template>
