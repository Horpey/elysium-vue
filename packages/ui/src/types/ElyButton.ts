export const ELY_BUTTON_THEME = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  DESTRUCTIVE: 'destructive',
} as const

export type ElyButtonThemeKey = keyof typeof ELY_BUTTON_THEME

export const ELY_BUTTON_SIZE = {
  LARGE: 'large',
  MEDIUM: 'medium',
  SMALL: 'small',
} as const

export type ElyButtonSizeKey = keyof typeof ELY_BUTTON_SIZE

export const ELY_BUTTON_TYPE = {
  BUTTON: 'button',
  SUBMIT: 'submit',
} as const

export type ElyButtonTypeKey = keyof typeof ELY_BUTTON_TYPE

export const ELY_BUTTON_TARGET = {
  BLANK: '_blank',
  PARENT: '_parent',
  SELF: '_self',
  TOP: '_top',
} as const

export type ElyButtonTargetKey = keyof typeof ELY_BUTTON_TARGET
