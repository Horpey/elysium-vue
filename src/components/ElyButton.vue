<script setup lang="ts">
interface Props {
  to?: string
  href?: string
  target?: string
  rel?: string
  type?: string
  disabled?: boolean
  onClick?: () => void
}

withDefaults(defineProps<Props>(), {
  to: undefined,
  href: undefined,
  target: undefined,
  rel: undefined,
  type: 'button',
  disabled: false,
  onClick: () => {},
})
</script>

<template>
  <RouterLink
    v-if="to"

    class="border px-5 py-3 text-xs font-light uppercase tracking-wider text-white shadow-btn transition-shadow duration-200 ease-in-out hover:shadow-none"
    :class="{
      'cursor-not-allowed opacity-50': disabled,
    }"
    :disabled="disabled"
    :to="to"
  >
    <slot />
  </RouterLink>

  <a
    v-else-if="href"
    class="border px-5 py-3 text-xs font-light uppercase tracking-wider text-white shadow-btn transition-shadow duration-200 ease-in-out hover:shadow-none"
    :class="{
      'cursor-not-allowed opacity-50': disabled,
    }"
    :disabled="disabled"
    :href="href"
    :target="target"
    :rel="rel"
    @click="onClick"
  >
    <slot />
  </a>

  <button
    v-else
    class="border px-5 py-3 text-xs font-light uppercase tracking-wider text-white shadow-btn transition-shadow duration-200 ease-in-out hover:shadow-none"
    :class="{
      'cursor-not-allowed opacity-50': disabled,
    }"
    :disabled="disabled"
    @click="onClick"
  >
    <slot />
  </button>
</template>
