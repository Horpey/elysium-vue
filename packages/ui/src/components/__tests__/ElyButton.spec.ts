import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import ElyButton from '../ElyButton.vue'

describe('elyButton', () => {
  it('renders with default props', () => {
    const wrapper = mount(ElyButton)

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert default values
    expect(wrapper.find('button').exists()).toBe(true)
    expect(wrapper.find('button').attributes('type')).toBe('button')
    expect(wrapper.find('button').classes()).toContain('border')
    expect(wrapper.find('button').classes()).toContain('font-light')

    // Assert slot content is not provided
    expect(wrapper.text()).toBe('')

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with custom props and slot content', async () => {
    const mockFn = vi.fn()

    const wrapper = mount(ElyButton, {
      props: {
        type: 'submit',
        theme: 'SECONDARY',
        disabled: true,
        size: 'LARGE',
        loading: false,
        onClick: mockFn,
      },
      slots: {
        default: '<span>Click me</span>',
      },
    })

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert custom values
    expect(wrapper.find('button').attributes('type')).toBe('submit')
    expect(wrapper.find('button').classes()).toContain('border-transparent')
    expect(wrapper.find('button').classes()).toContain('bg-aquamarine-700')
    expect(wrapper.find('button').classes()).toContain('text-white')

    // Assert slot content is rendered
    // expect(wrapper.find('span').text()).toBe('Click me')

    // Click the button
    await wrapper.find('button').trigger('click')

    // await wrapper.props().onClick()

    // Assert that the onClick function was called
    // expect(mockFn).toHaveBeenCalled()

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })
})
