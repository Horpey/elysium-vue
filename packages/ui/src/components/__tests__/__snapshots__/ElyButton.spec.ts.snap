// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`elyButton > renders with custom props and slot content 1`] = `
"<button class="border font-light uppercase tracking-wider transition-all duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-opacity-50 border-transparent bg-aquamarine-700 text-white focus:ring-aquamarine-600 hover:bg-aquamarine-600 hover:text-white text-base px-6 py-3 cursor-not-allowed opacity-50" disabled="" type="submit"><span class=""><!--v-if--><!--v-if--><span><span>Click me</span></span>
  <!--v-if--></span>
</button>"
`;

exports[`elyButton > renders with default props 1`] = `
"<button class="border font-light uppercase tracking-wider transition-all duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-opacity-50 border-transparent bg-koromiko-700 text-white focus:ring-koromiko-600 hover:bg-koromiko-600 hover:text-white text-sm px-4 py-2" type="button"><span class="space-x-2"><!--v-if--><!--v-if--><span></span>
  <!--v-if--></span>
</button>"
`;
